### PR TITLE
Button based secondary attack feature

### PR DESCRIPTION
**Pull request by Aceship**
Add Button based Secondary attack
- Hold Grab + trigger to start secondary attack 
- release the trigger to end the secondary attack

when the the secondary ends, when it hits enemy, it will damage the enemy on some delay

- when using other than atgeir, it should attack the enemy that this around the slash effect (note that most weapon doesnt allow multi hit, but currently this allow multi hit when some enemy is near enough)

- when using atgeir, it will show a slash around the character when secondary attack is used, and it will attack 360 degrees around the character based on the height where you do the secondary attack

also adds vanilla-like pickaxe behavior by using this secondary attack

### CODE REVIEW

File: ValheimVRMod/Scripts/WeaponSecondaryManager.cs [hunk 1]
```diff
@@ -0,0 +1,571 @@
...
+            if (!SteamVR_Actions.valheim_Grab.GetState(mainHandSource) || 
+                item.m_shared.m_itemType == ItemDrop.ItemData.ItemType.TwoHandedWeapon && !WeaponWield.isCurrentlyTwoHanded())
+            {
+                firstPos = Vector3.zero;
+                lastPos = Vector3.zero;
+            }
+            
+            if (SteamVR_Actions.valheim_Grab.GetState(mainHandSource) && 
+                !inCooldown && 
+                !VRPlayer.vrPlayerInstance.CheckMenuIsOpen() &&
```
Comment (hhliow) [inline ValheimVRMod/Scripts/WeaponSecondaryManager.cs:224]: Do we need allow secondary attack here in the following scenario?
1. Console is visible
2. Chatting
3. Player is in cutscene

File: ValheimVRMod/Patches/EquipPatches.cs [hunk 2]
```diff
@@ -313,4 +315,23 @@ void OnRenderObject()
...
+    {
+        static bool Prefix(Player __instance, ref bool __result)
+        {
+            if(__instance != Player.m_localPlayer || !VHVRConfig.UseVrControls())
+            {
+                return true;
+            }
+
+            if (WeaponSecondaryManager.wasSecondaryAttack)
+            {
```
Comment (hhliow) [inline ValheimVRMod/Patches/EquipPatches.cs:330]: What's the purpose of this change?

File: ValheimVRMod/Scripts/WeaponSecondaryManager.cs [hunk 3]
```diff
@@ -0,0 +1,571 @@
...
+using UnityEngine;
+using UnityEngine.Rendering;
+using ValheimVRMod.Scripts.Block;
+using ValheimVRMod.Utilities;
+using ValheimVRMod.VRCore;
+using Valve.VR;
+
+namespace ValheimVRMod.Scripts
+{
+    class WeaponSecondaryManager : MonoBehaviour
```
Comment (hhliow) [inline ValheimVRMod/Scripts/WeaponSecondaryManager.cs:14]: Consider renaming to ButtonSecondaryAttackManager

File: ValheimVRMod/Patches/CollisionPatches.cs [hunk 4]
```diff
@@ -177,7 +177,15 @@ class PatchAttackStart {
                 float randomSkillFactor = ___m_character.GetRandomSkillFactor(skill);
 
                 if (___m_lowerDamagePerHit) {
-                    randomSkillFactor /= 0.75f;
+                    
+                    if(WeaponSecondaryManager.wasSecondaryAttack && WeaponSecondaryManager.secondaryHitList.Count >= 1)
+                    {
+                        randomSkillFactor /= (WeaponSecondaryManager.secondaryHitList.Count - WeaponSecondaryManager.hitExtra) * 0.75f;
```
Comment (hhliow) [inline ValheimVRMod/Patches/CollisionPatches.cs:183]: What does hitExtra do exactly?

File: ValheimVRMod/Patches/EquipPatches.cs (hunk 2 above)
Comment (Aceship) [reply ValheimVRMod/Patches/EquipPatches.cs:330]: so you cant change weapon mid secondary attack cooldown

File: ValheimVRMod/Patches/CollisionPatches.cs (hunk 4 above)
Comment (Aceship) [reply ValheimVRMod/Patches/CollisionPatches.cs:183]: well technically it so just attack that hits terrain doesnt reduce more damage

File: ValheimVRMod/Scripts/WeaponSecondaryManager.cs (hunk 1 above)
Comment (Aceship) [reply ValheimVRMod/Scripts/WeaponSecondaryManager.cs:224]: ngl at first, its just because you can accidentally hit stuff when opening chest and inventory

while chatting probably you cant activate it when using steam vr keyboard

console probably should still work 

for cutscene, probably not, but cutscene is just when starting the game and flying with the bird ?
you dont have any weapon on that time anyway so i guess its okay ?

File: ValheimVRMod/Scripts/WeaponSecondaryManager.cs (hunk 3 above)
Comment (Aceship) [reply ValheimVRMod/Scripts/WeaponSecondaryManager.cs]: Done 👍

File: ValheimVRMod/Patches/EquipPatches.cs (hunk 2 above)
Comment (hhliow) [reply ValheimVRMod/Patches/EquipPatches.cs:330]: In all capacity this variable could probably be renamed isSecondaryAttackTriggered or isSecondaryAttackPending or something along the line

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs [hunk 5]
```diff
@@ -0,0 +1,571 @@
...
+                    foreach (var hit in secondaryHitList)
+                    {
+                        var target = hit.collider.gameObject;
+                        if (target.GetComponent<Heightmap>() == null)
+                        {
+                            isTerrain = false;
+                        }
+                        else
+                        {
+                            hitExtra += 1;
```
Comment (hhliow) [inline ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs:439]: let's rename this to terrainHitCount

File: ValheimVRMod/Patches/CollisionPatches.cs [hunk 6]
```diff
@@ -177,7 +177,15 @@ class PatchAttackStart {
                 float randomSkillFactor = ___m_character.GetRandomSkillFactor(skill);
 
                 if (___m_lowerDamagePerHit) {
-                    randomSkillFactor /= 0.75f;
+                    
+                    if(ButtonSecondaryAttackManager.wasSecondaryAttack && ButtonSecondaryAttackManager.secondaryHitList.Count >= 1)
+                    {
+                        randomSkillFactor /= (ButtonSecondaryAttackManager.secondaryHitList.Count - ButtonSecondaryAttackManager.hitExtra) * 0.75f;
```
Comment (hhliow) [inline ValheimVRMod/Patches/CollisionPatches.cs:183]: Consider adding a method or getter ButtonSecondaryAttackManager.nonTerrainHitCount

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs [hunk 7]
```diff
@@ -0,0 +1,571 @@
...
+        {
+            firstPos = Vector3.zero;
+            lastPos = Vector3.zero;
+            secondaryHitList = new List<Attack.HitPoint>();
+            slashLine.enabled = false;
+            wasSecondaryAttack = false;
+            wasSecondaryAttacked = true;
+        }
+
+        public void Initialize(Transform obj, string name, bool rightHand)
```
Comment (hhliow) [inline ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs:107]: isRightHand

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs [hunk 8]
```diff
@@ -0,0 +1,571 @@
...
+        {
+            parent = obj.parent.gameObject;
+            outline = obj.parent.gameObject.GetComponent<Outline>();
+            isRightHand = rightHand;
```
Comment (hhliow) [inline ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs:111]: this.isRightHand = isRightHand

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs [hunk 9]
```diff
@@ -0,0 +1,571 @@
...
+            {
+                return;
+            }
+
+            if (secondaryAttackTimer >= secondaryAttackTimerFull)
+            {
+                secondaryAttackTimer -= Time.deltaTime;
+            }
+
+            var mainHand = isRightHand ? VRPlayer.rightHand : VRPlayer.leftHand;
```
Comment (hhliow) [inline ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs:202]: You remove this and use VRPlayer.dominantHand instead

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs [hunk 10]
```diff
@@ -0,0 +1,571 @@
...
+            var offHand = isRightHand ? VRPlayer.leftHand : VRPlayer.rightHand;
```
Comment (hhliow) [inline ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs:203]: similarly VRPlayer.nonDominantHand

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs [hunk 11]
```diff
@@ -0,0 +1,571 @@
...
+            var mainHandSource = isRightHand ? SteamVR_Input_Sources.RightHand : SteamVR_Input_Sources.LeftHand;
```
Comment (hhliow) [inline ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs:204]: VRPlayer.mainHandInputSource

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs [hunk 12]
```diff
@@ -0,0 +1,571 @@
...
+            if (!SteamVR_Actions.valheim_Grab.GetState(mainHandSource) || 
+                item.m_shared.m_itemType == ItemDrop.ItemData.ItemType.TwoHandedWeapon && !WeaponWield.isCurrentlyTwoHanded())
+            {
+                firstPos = Vector3.zero;
+                lastPos = Vector3.zero;
+            }
+            
+            if (SteamVR_Actions.valheim_Grab.GetState(mainHandSource) && 
+                !inCooldown && 
+                !VRPlayer.vrPlayerInstance.CheckMenuIsOpen() &&
```
Comment (hhliow) [inline ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs:224]: isClickableGuiOpen

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs [hunk 13]
```diff
@@ -0,0 +1,571 @@
...
+
+            if (firstPos != Vector3.zero && lastPos == Vector3.zero)
+            {
+                if (secondaryAttack.m_attackAnimation == "atgeir_secondary")
+                {
+                    var currpos = localHandPos + localWeaponForward;
+                    var angle = Vector3.SignedAngle(firstPos, currpos, Player.m_localPlayer.transform.up);
+                    var anglestep = angle / 2;
+                    var positive = angle >= 0 ? 1 : -1;
+                    slashLine.positionCount = ((int)angle * positive) / 4;
```
Comment (hhliow) [inline ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs:295]: Math.Abs(angle)

**Review by hhliow — COMMENTED** [review]
Thanks!

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs (hunk 10 above)
Comment (Aceship) [reply ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs:203]: this doesnt seems to exist ?
tho i can use VRPlayer.dominantHand.otherHand i think

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs (hunk 5 above)
Comment (Aceship) [reply ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs]: Done 👍

File: ValheimVRMod/Patches/EquipPatches.cs (hunk 2 above)
Comment (Aceship) [reply ValheimVRMod/Patches/EquipPatches.cs:330]: renamed the variable to `isSecondaryAttackStarted`
since there's also `wasSecondaryAttacked` renamed to  `isSecondaryAttackTriggered` where the attack just triggered on the enemy
also changed `secondaryAttackJustEnded` to `isSecondaryAttackEnded` (this one is for vibration to make it easier to know when the secondary attack slowdown finished)

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs (hunk 7 above)
Comment (Aceship) [reply ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs]: Done 👍

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs (hunk 8 above)
Comment (Aceship) [reply ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs]: Done 👍

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs (hunk 9 above)
Comment (Aceship) [reply ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs]: Done 👍

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs (hunk 11 above)
Comment (Aceship) [reply ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs]: there's no `VRPlayer.mainHandInputSource` but there's `VRPlayer.dominantHandInputSource` so i used that one

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs (hunk 12 above)
Comment (Aceship) [reply ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs]: Done 👍

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs (hunk 13 above)
Comment (Aceship) [reply ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs]: Done 👍

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs [hunk 14]
```diff
@@ -0,0 +1,571 @@
...
+            
+        }
+        private void Update()
+        {
+            if (!outline)
+            {
+                outline = parent.GetComponent<Outline>();
+                return;
+            }
+            SecondaryAttack();
```
Comment (hhliow) [inline ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs:190]: This method contains the bulk of the logic existing in this class, ideally it could be broken down to a few methods reflecting the major steps for checking secondary attack each frame.

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs [hunk 15]
```diff
@@ -0,0 +1,571 @@
...
+            if (!(inCooldown || firstPos != Vector3.zero))
+            {
+                if (secondaryAttackTimer <= secondaryAttackTimerFull)
+                {
+                    isSecondaryAttackStarted = false;
+                    slashTrail.emitting = false;
+                    outline.enabled = false;
+                    if (!isSecondaryAttackEnded)
+                    {
+                        if (isRightHand)
```
Comment (hhliow) [inline ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs:269]: Could do VRPlayer.dominantHand.hapticAction.Execute(... VRPlayer.dominantHandInputSource)

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs [hunk 16]
```diff
@@ -0,0 +1,571 @@
...
+                GameObject gameObject = Projectile.FindHitObject(raycastHit.collider);
+                if (gameObject && !(gameObject == Player.m_localPlayer.gameObject))
+                {
+                    Vagon component = gameObject.GetComponent<Vagon>();
+                    if (!component || !component.IsAttached(Player.m_localPlayer))
+                    {
+                        Character component2 = gameObject.GetComponent<Character>();
+                        if (component2 != null)
+                        {
+                            bool flag = BaseAI.IsEnemy(Player.m_localPlayer, component2) || (component2.GetBaseAI() && component2.GetBaseAI().IsAggravatable() && Player.m_localPlayer.IsPlayer());
```
Comment (hhliow) [inline ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs:536]: If we have going into repro detailed vanilla-like logic like this, I wonder if we should consider just play the vanilla animation and let vanilla game handle the attack instead. Although I know in this case that vanilla atgeir swipe doesn't have height variation whereas we attempt consider height.

**Review by hhliow — COMMENTED** [review]
Thanks!

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs (hunk 15 above)
Comment (Aceship) [reply ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs]: Done 👍

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs (hunk 16 above)
Comment (Aceship) [reply ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs:529]: i dont think we can actually do that because the logic is in the middle of `DoMeleeAttack` and need to modify a lot of stuff to actually make it work, it also have a more hardcoded swing direction and angle

using it means we can only make it hit on a specific area in front of the character

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs [hunk 17]
```diff
@@ -0,0 +1,564 @@
...
+                return;
+            }
+
+            foreach (var raycastHit in raycastList)
+            {
+                if (raycastHit.collider.gameObject == Player.m_localPlayer.gameObject)
+                {
+                    continue;
+                }
+                GameObject gameObject = Projectile.FindHitObject(raycastHit.collider);
```
Comment (hhliow) [inline ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs:520]: hitObject

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs [hunk 18]
```diff
@@ -0,0 +1,564 @@
...
+                if (gameObject && !(gameObject == Player.m_localPlayer.gameObject))
```
Comment (hhliow) [inline ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs:521]: !=

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs [hunk 19]
```diff
@@ -0,0 +1,564 @@
...
+                {
+                    Vagon component = gameObject.GetComponent<Vagon>();
```
Comment (hhliow) [inline ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs:523]: Vagon vagon = ...

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs [hunk 20]
```diff
@@ -0,0 +1,564 @@
...
+                    if (!component || !component.IsAttached(Player.m_localPlayer))
+                    {
+                        Character component2 = gameObject.GetComponent<Character>();
```
Comment (hhliow) [inline ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs:526]: targetCharacter = ...

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs [hunk 21]
```diff
@@ -0,0 +1,564 @@
...
+                        if (component2 != null)
+                        {
+                            bool flag = BaseAI.IsEnemy(Player.m_localPlayer, component2) || (component2.GetBaseAI() && component2.GetBaseAI().IsAggravatable() && Player.m_localPlayer.IsPlayer());
```
Comment (hhliow) [inline ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs:529]: What does Player.m_localPlayer.IsPlayer() check?

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs [hunk 22]
```diff
@@ -0,0 +1,564 @@
...
+                            if ((!Player.m_localPlayer.IsPlayer() && !flag) || (!item.m_shared.m_tamedOnly && Player.m_localPlayer.IsPlayer() && !Player.m_localPlayer.IsPVPEnabled() && !flag) || (item.m_shared.m_tamedOnly && !component2.IsTamed()))
```
Comment (hhliow) [inline ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs:530]: !Player.m_localPlayer.IsPlayer() && !flag
= !Player.m_localPlayer.IsPlayer() && !(BaseAI.IsEnemy(Player.m_localPlayer, component2) || (component2.GetBaseAI() && component2.GetBaseAI().IsAggravatable() && Player.m_localPlayer.IsPlayer()))
= !Player.m_localPlayer.IsPlayer() && !BaseAI.IsEnemy(Player.m_localPlayer, component2) && !(component2.GetBaseAI() && component2.GetBaseAI().IsAggravatable() && Player.m_localPlayer.IsPlayer()))
= !Player.m_localPlayer.IsPlayer() && !BaseAI.IsEnemy(Player.m_localPlayer, component2) && (!component2.GetBaseAI() || !component2.GetBaseAI().IsAggravatable() || !Player.m_localPlayer.IsPlayer()))


since !Player.m_localPlayer.IsPlayer() implies (!component2.GetBaseAI() || !component2.GetBaseAI().IsAggravatable() || !Player.m_localPlayer.IsPlayer())),

the above is just equal to
!Player.m_localPlayer.IsPlayer() && !BaseAI.IsEnemy(Player.m_localPlayer, component2)

There is no need to check for (component2.GetBaseAI() && component2.GetBaseAI().IsAggravatable())

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs (hunk 21 above)
Comment (Aceship) [reply ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs:529]: ngl mostly on this part is just copied from the vanilla but just replaced the `this.m_character.IsPlayer()` to that, so its just mostly vanilla logic

Comment (hhliow) [reply ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs:529]: double check but i suspect Player.m_localPlayer.IsPlayer() always returns true anyway

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs (hunk 22 above)
Comment (Aceship) [reply ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs:530]: > There is no need to check for (component2.GetBaseAI() && component2.GetBaseAI().IsAggravatable())

from what i tried, thats actually used for checking if you hitting friendly like dvergr on mistland, since usually its friendly and you can hit it to make it aggro to you, without that part, you cant hit it unless its aggro to you 

i guess i leave only that part alone so it works against friendly that can be aggroed

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs (hunk 21 above)
Comment (Aceship) [reply ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs]: seems like its supposed to be the attacker, but since we used our own attack, it never register the character to the `attack.m_character`

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs (hunk 20 above)
Comment (Aceship) [reply ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs]: done 👍

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs (hunk 19 above)
Comment (Aceship) [reply ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs]: done 👍

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs (hunk 18 above)
Comment (Aceship) [reply ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs]: done 👍

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs (hunk 17 above)
Comment (Aceship) [reply ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs]: done 👍

File: ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs [hunk 23]
```diff
@@ -0,0 +1,564 @@
...
+                if (hitObject && hitObject != Player.m_localPlayer.gameObject)
+                {
+                    Vagon vagon = hitObject.GetComponent<Vagon>();
+                    if (!vagon || !vagon.IsAttached(Player.m_localPlayer))
+                    {
+                        Character targetCharacter = hitObject.GetComponent<Character>();
+                        if (targetCharacter != null)
+                        {
+                            bool flag = BaseAI.IsEnemy(Player.m_localPlayer, targetCharacter) || (targetCharacter.GetBaseAI() && targetCharacter.GetBaseAI().IsAggravatable());
+                            if ((!flag) || (!item.m_shared.m_tamedOnly && !Player.m_localPlayer.IsPVPEnabled() && !flag) || (item.m_shared.m_tamedOnly && !targetCharacter.IsTamed()))
```
Comment (hhliow) [inline ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs:530]: since Player.m_localPlayer.isLocalPlayer() evaluates to true,
(!Player.m_localPlayer.isLocalPlayer() && !flag) evaluates to false, not necessarily !flag
Where are the vanilla code for this found and how to access them? Might worth add a comment for that if the intention is to replace this with new vanilla logic as they changed.

Comment (Aceship) [reply ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs:530]: technically its supposed to not hit when the target is not enemy and not aggro able , so isnt it supposed to do that ?

its on `Attack.DoMeleeAttack()`

the original code is this 
image

Comment (Aceship) [reply ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs:530]: hmm i see, i guess its still always supposedly false on that `(!flag)`

Comment (hhliow) [reply ValheimVRMod/Scripts/ButtonSecondaryAttackManager.cs:530]: That check is only for non-player attackers. Here since the attacker is the player it always evaluates to false.

**Review by hhliow — APPROVED** [review]
